src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java

```diff
@@ -534,109 +544,215 @@ public void force(boolean metaData) throws IOException {
         }
     }
 
-    // Assume at first that the underlying kernel supports sendfile();
+    // Assume at first that the underlying kernel supports sendfile/equivalent;
     // set this to true if we find out later that it doesn't
     //
-    private static volatile boolean transferToNotSupported;
+    private static volatile boolean transferToDirectNotSupported;
 
-    // Assume that the underlying kernel sendfile() will work if the target
-    // fd is a pipe; set this to false if we find out later that it doesn't
+    // Assume that the underlying kernel sendfile/equivalent will work if the target
+    // fd is a file; set this to true if we find out later that it doesn't
     //
-    private static volatile boolean pipeSupported = true;
+    private static volatile boolean transferToFileDirectNotSupported;
 
-    // Assume that the underlying kernel sendfile() will work if the target
-    // fd is a file; set this to false if we find out later that it doesn't
-    //
-    private static volatile boolean fileSupported = true;
+    /**
+     * Marks the beginning of a transfer to or from this channel.
+     * @throws ClosedChannelException if channel is closed
+     */
+    private int beforeTransfer() throws ClosedChannelException {
```

Comment:
Should this method and `afterTransfer()` be moved up before the `*NotSupported` fields as they cover transferring both to and from?
```
    // beforeTransfer()
    // afterTransfer()
    // transferTo fields and methods
    // transferFrom fields and methods
```

Reply:
Alternatively, transferFromNotSupported; can be moved up so that the transfer flags are in one place.

Reply:
Either way.